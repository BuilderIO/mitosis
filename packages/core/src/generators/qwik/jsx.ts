import { MitosisNode } from '../../types/mitosis-node';
import { DIRECTIVES } from './directives';
import {
  File,
  INDENT,
  invoke,
  NL,
  SrcBuilder,
  quote,
  UNINDENT,
  lastProperty,
} from './src-generator';
import { CssStyles } from './styles';

export function renderJSXNodes(
  file: File,
  directives: Map<string, string>,
  handlers: Map<string, string>,
  children: MitosisNode[],
  styles: Map<string, CssStyles>,
  parentSymbolBindings: Record<string, string>,
  root = true,
): any {
  return function(this: SrcBuilder) {
    if (children.length == 0) return;
    if (root) this.emit('(', INDENT, NL);
    const needsFragment = root && children.length > 1;
    file.import(file.qwikModule, 'h');
    if (needsFragment) {
      file.import(file.qwikModule, 'Fragment');
      this.jsxBeginFragment(file.import(file.qwikModule, 'Fragment'));
    }
    children.forEach((child) => {
      if (isEmptyTextNode(child)) return;
      if (isTextNode(child)) {
        if (child.bindings._text !== undefined) {
          this.jsxTextBinding(child.bindings._text);
        } else {
          this.isJSX
            ? this.emit(child.properties._text)
            : this.jsxTextBinding(quote(child.properties._text!));
        }
      } else {
        let childName = child.name;
        const directive = DIRECTIVES[childName];
        if (typeof directive == 'function') {
          this.emit(
            directive(child, () =>
              renderJSXNodes(
                file,
                directives,
                handlers,
                child.children,
                styles,
                {},
                false,
              ).call(this),
            ),
          );
        } else {
          if (typeof directive == 'string') {
            directives.set(childName, directive);
            if (file.module !== 'med') {
              file.import('./med.js', childName);
            }
          }
          if (isSymbol(childName)) {
            // TODO(misko): We are hard coding './med.js' which is not right.
            file.import('./med.js', childName);
            let exportedChildName = file.exports.get(childName);
            if (exportedChildName) {
              childName = exportedChildName;
            }
          }
          let props: Record<string, any> = child.properties;
          const css = child.bindings.css;
          if (css) {
            props = { ...props };
            props.class = addClass(styles.get(css)!.CLASS_NAME, props.class);
          }
          const symbolBindings: Record<string, string> = {};
          const bindings = rewriteHandlers(
            file,
            handlers,
            child.bindings,
            symbolBindings,
          );
          this.jsxBegin(childName, props, {
            ...bindings,
            ...parentSymbolBindings,
          });
          renderJSXNodes(
            file,
            directives,
            handlers,
            child.children,
            styles,
            symbolBindings,
            false,
          ).call(this);
          this.jsxEnd(childName);
        }
      }
    });
    if (needsFragment) {
      this.jsxEndFragment();
    }
    if (root) this.emit(UNINDENT, ')');
  };
}

function isSymbol(name: string): boolean {
  return name.charAt(0) == name.charAt(0).toUpperCase();
}

function addClass(
  className: string,
  existingClass: string | undefined,
): string {
  return [className, ...(existingClass ? existingClass.split(' ') : [])].join(
    ' ',
  );
}

function isEmptyTextNode(child: MitosisNode) {
  return child.properties._text?.trim() == '';
}

function isTextNode(child: MitosisNode) {
  return (
    child.properties._text !== undefined || child.bindings._text !== undefined
  );
}

/**
 * Rewrites bindings:
 * - Remove `css`
 * - Rewrites event handles
 * - Extracts symbol bindings.
 *
 * @param file
 * @param handlers
 * @param bindings
 * @param symbolBindings Options record which will receive the symbol bindings
 * @returns
 */
function rewriteHandlers(
  file: File,
  handlers: Map<string, string>,
  bindings: Record<string, string | undefined>,
  symbolBindings: Record<string, string>,
): Record<string, string> {
  const outBindings: Record<string, string> = {};
  for (let key in bindings) {
    if (Object.prototype.hasOwnProperty.call(bindings, key)) {
      let binding: any = bindings[key]!;
      let handlerBlock: string | undefined;
      if (binding != null) {
        if (key == 'css') {
          continue;
        } else if ((handlerBlock = handlers.get(binding))) {
          key = `on:${key.substring(2).toLowerCase()}`;
          binding = invoke(file.import(file.qwikModule, 'qrl'), [
            quote(file.qrlPrefix + 'high.js'),
            quote(handlerBlock),
            '[__props__, __state__]',
          ]);
        } else if (symbolBindings && key.startsWith('symbol.data.')) {
          symbolBindings[lastProperty(key)] = binding;
        }
        outBindings[key] = binding;
      }
    }
  }
  return outBindings;
}
