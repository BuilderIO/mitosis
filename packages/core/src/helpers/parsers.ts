import * as babel from '@babel/core';

const jsxPlugin = require('@babel/plugin-syntax-jsx');
const tsPreset = require('@babel/preset-typescript');
const decorators = require('@babel/plugin-syntax-decorators');

export function parseCode(code: string) {
  const ast = babel.parse(code, {
    presets: [[tsPreset, { isTSX: true, allExtensions: true }]],
    plugins: [[decorators, { legacy: true }], jsxPlugin],
  });
  const body = babel.types.isFile(ast)
    ? ast.program.body
    : babel.types.isProgram(ast)
    ? ast.body
    : [];
  return body;
}

export const isCodeBodyExpression = (body: babel.types.Statement[]) =>
  body.length == 1 &&
  (babel.types.isExpression(body[0]) ||
    babel.types.isExpressionStatement(body[0]));

/**
 * Returns `true` if the `code` is a valid expression. (vs a statement)
 */
export function isExpression(code: string): boolean {
  try {
    const body = parseCode(code);
    return isCodeBodyExpression(body);
  } catch (e) {
    return false;
  }
}

export const isCodeBodyIdentifier = (body: babel.types.Statement[]) =>
  body.length == 1 && babel.types.isIdentifier(body[0]);
