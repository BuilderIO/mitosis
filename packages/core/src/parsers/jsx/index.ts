export * from './jsx';
export { selfClosingTags } from './helpers';
export { METADATA_HOOK_NAME } from './hooks';
