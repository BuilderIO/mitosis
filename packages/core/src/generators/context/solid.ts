import { format } from 'prettier/standalone';
import { getMemberObjectString } from '../../helpers/get-state-object-string';
import { MitosisContext } from '../../types/mitosis-context';

type ContextToSolidOptions = {
  format?: boolean;
};

export const contextToSolid =
  (options: ContextToSolidOptions = {}) =>
  ({ context }: { context: MitosisContext }): string => {
    let str = `
    import { createContext } from 'solid-js';

    export default createContext(${getMemberObjectString(context.value)})
    `;

    if (options.format !== false) {
      try {
        str = format(str, {
          parser: 'typescript',
          plugins: [
            require('prettier/parser-typescript'), // To support running in browsers
          ],
        });
      } catch (err) {
        console.error('Format error for file:', str);
        throw err;
      }
    }

    return str;
  };
