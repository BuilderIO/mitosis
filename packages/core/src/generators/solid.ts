import dedent from 'dedent';
import { format } from 'prettier/standalone';
import { hasStyles } from '../helpers/collect-styles';
import { getRefs } from '../helpers/get-refs';
import {
  getMemberObjectString,
  getStateObjectStringFromComponent,
} from '../helpers/get-state-object-string';
import { renderPreComponent } from '../helpers/render-imports';
import { selfClosingTags } from '../parsers/jsx';
import { MitosisComponent } from '../types/mitosis-component';
import { MitosisNode } from '../types/mitosis-node';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../modules/plugins';
import { fastClone } from '../helpers/fast-clone';
import { stripMetaProperties } from '../helpers/strip-meta-properties';
import { getComponentsUsed } from '../helpers/get-components-used';
import traverse from 'traverse';
import { isMitosisNode } from '../helpers/is-mitosis-node';
import { BaseTranspilerOptions, Transpiler } from '../types/config';
import { filterEmptyTextNodes } from '../helpers/filter-empty-text-nodes';
import { createMitosisNode } from '../helpers/create-mitosis-node';
import { hasContext } from './helpers/context';
import { babelTransformExpression } from '../helpers/babel-transform';
import { types } from '@babel/core';
import { kebabCase } from 'lodash';

export interface ToSolidOptions extends BaseTranspilerOptions {}

// Transform <foo.bar key="value" /> to <component :is="foo.bar" key="value" />
function processDynamicComponents(
  json: MitosisComponent,
  options: ToSolidOptions,
) {
  let found = false;
  traverse(json).forEach((node) => {
    if (isMitosisNode(node)) {
      if (node.name.includes('.')) {
        (node.bindings.component as { code: string }).code = node.name;
        node.name = 'Dynamic';
        found = true;
      }
    }
  });
  return found;
}

function getContextString(
  component: MitosisComponent,
  options: ToSolidOptions,
) {
  let str = '';
  for (const key in component.context.get) {
    str += `
      const ${key} = useContext(${component.context.get[key].name});
    `;
  }

  return str;
}

// This should really be a preprocessor mapping the `class` attribute binding based on what other values have
// to make this more pluggable
const collectClassString = (json: MitosisNode): string | null => {
  const staticClasses: string[] = [];

  const hasStaticClasses = Boolean(staticClasses.length);
  if (json.properties.class) {
    staticClasses.push(json.properties.class);
    delete json.properties.class;
  }
  if (json.properties.className) {
    staticClasses.push(json.properties.className);
    delete json.properties.className;
  }

  const dynamicClasses: string[] = [];
  if (typeof json.bindings.class?.code === 'string') {
    dynamicClasses.push(json.bindings.class.code as any);
    delete json.bindings.class;
  }
  if (typeof json.bindings.className?.code === 'string') {
    dynamicClasses.push(json.bindings.className.code as any);
    delete json.bindings.className;
  }
  if (
    typeof json.bindings.css?.code === 'string' &&
    json.bindings.css.code.trim().length > 4
  ) {
    dynamicClasses.push(`css(${json.bindings.css.code})`);
  }
  delete json.bindings.css;
  const staticClassesString = staticClasses.join(' ');

  const dynamicClassesString = dynamicClasses.join(" + ' ' + ");

  const hasDynamicClasses = Boolean(dynamicClasses.length);

  if (hasStaticClasses && !hasDynamicClasses) {
    return `"${staticClassesString}"`;
  }

  if (hasDynamicClasses && !hasStaticClasses) {
    return `{${dynamicClassesString}}`;
  }

  if (hasDynamicClasses && hasStaticClasses) {
    return `{"${staticClassesString} " + ${dynamicClassesString}}`;
  }

  return null;
};

const blockToSolid = (
  json: MitosisNode,
  options: ToSolidOptions = {},
): string => {
  if (json.properties._text) {
    return json.properties._text;
  }
  if (json.bindings._text?.code) {
    return `{${json.bindings._text.code}}`;
  }

  if (json.name === 'For') {
    const needsWrapper = json.children.length !== 1;
    // The SolidJS `<For>` component has a special index() signal function.
    // https://www.solidjs.com/docs/latest#%3Cfor%3E
    return `<For each={${json.bindings.each?.code}}>
    {(${json.properties._forName}, _index) => {
      const index = _index();
      return ${needsWrapper ? '<>' : ''}${json.children
      .filter(filterEmptyTextNodes)
      .map((child) => blockToSolid(child, options))}}}
      ${needsWrapper ? '</>' : ''}
    </For>`;
  }

  let str = '';

  if (json.name === 'Fragment') {
    str += '<';
  } else {
    str += `<${json.name} `;
  }

  if (json.name === 'Show' && json.meta.else) {
    str += `fallback={${blockToSolid(json.meta.else as any, options)}}`;
  }

  const classString = collectClassString(json);
  if (classString) {
    str += ` class=${classString} `;
  }

  if (json.bindings._spread?.code) {
    str += ` {...(${json.bindings._spread.code})} `;
  }

  for (const key in json.properties) {
    const value = json.properties[key];
    str += ` ${key}="${value}" `;
  }
  for (const key in json.bindings) {
    const value = json.bindings[key];
    if (key === '_spread' || key === '_forName') {
      continue;
    }
    if (!value?.code) continue;

    if (key.startsWith('on')) {
      const useKey =
        key === 'onChange' && json.name === 'input' ? 'onInput' : key;
      str += ` ${useKey}={event => ${value.code}} `;
    } else {
      let useValue = value.code;
      if (key === 'style') {
        // Convert camelCase keys to kebab-case
        // TODO: support more than top level objects, may need
        // a runtime helper for expressions that are not a direct
        // object literal, such as ternaries and other expression
        // types
        useValue = babelTransformExpression(value.code, {
          ObjectExpression(path: babel.NodePath<babel.types.ObjectExpression>) {
            // TODO: limit to top level objects only
            for (const property of path.node.properties) {
              if (types.isObjectProperty(property)) {
                if (
                  types.isIdentifier(property.key) ||
                  types.isStringLiteral(property.key)
                ) {
                  const key = types.isIdentifier(property.key)
                    ? property.key.name
                    : property.key.value;
                  property.key = types.stringLiteral(kebabCase(key));
                }
              }
            }
          },
        });
      }
      str += ` ${key}={${useValue}} `;
    }
  }
  if (selfClosingTags.has(json.name)) {
    return str + ' />';
  }
  str += '>';
  if (json.children) {
    str += json.children
      .filter(filterEmptyTextNodes)
      .map((item) => blockToSolid(item, options))
      .join('\n');
  }

  if (json.name === 'Fragment') {
    str += '</>';
  } else {
    str += `</${json.name}>`;
  }

  return str;
};

const getRefsString = (json: MitosisComponent, refs = getRefs(json)) => {
  let str = '';

  for (const ref of Array.from(refs)) {
    str += `\nconst ${ref} = useRef();`;
  }

  return str;
};

function addProviderComponents(
  json: MitosisComponent,
  options: ToSolidOptions,
) {
  for (const key in json.context.set) {
    const { name, value } = json.context.set[key];
    json.children = [
      createMitosisNode({
        name: `${name}.Provider`,
        children: json.children,
        ...(value && {
          bindings: {
            value: { code: getMemberObjectString(value) },
          },
        }),
      }),
    ];
  }
}

export const componentToSolid =
  (options: ToSolidOptions = {}): Transpiler =>
  ({ component }) => {
    let json = fastClone(component);
    if (options.plugins) {
      json = runPreJsonPlugins(json, options.plugins);
    }
    addProviderComponents(json, options);
    const componentHasStyles = hasStyles(json);
    const addWrapper = json.children.filter(filterEmptyTextNodes).length !== 1;
    if (options.plugins) {
      json = runPostJsonPlugins(json, options.plugins);
    }
    stripMetaProperties(json);
    const foundDynamicComponents = processDynamicComponents(json, options);

    const stateString = getStateObjectStringFromComponent(json);
    const hasState = Boolean(Object.keys(component.state).length);
    const componentsUsed = getComponentsUsed(json);
    const componentHasContext = hasContext(json);

    const hasShowComponent = componentsUsed.has('Show');
    const hasForComponent = componentsUsed.has('For');

    const solidJSImports = [
      componentHasContext ? 'useContext' : undefined,
      hasShowComponent ? 'Show' : undefined,
      hasForComponent ? 'For' : undefined,
      json.hooks.onMount?.code ? 'onMount' : undefined,
    ].filter(Boolean);

    let str = dedent`
    ${
      solidJSImports.length > 0
        ? `import { 
          ${solidJSImports.map((item) => item).join(', ')}
         } from 'solid-js';`
        : ''
    }
    ${!foundDynamicComponents ? '' : `import { Dynamic } from 'solid-js/web';`}
    ${!hasState ? '' : `import { createMutable } from 'solid-js/store';`}
    ${
      !componentHasStyles
        ? ''
        : `import { css } from "solid-styled-components";`
    }
    ${renderPreComponent(json)}

    function ${json.name}(props) {
      ${!hasState ? '' : `const state = createMutable(${stateString});`}
      
      ${getRefsString(json)}
      ${getContextString(json, options)}

      ${
        !json.hooks.onMount?.code
          ? ''
          : `onMount(() => { ${json.hooks.onMount.code} })`
      }

      return (${addWrapper ? '<>' : ''}
        ${json.children
          .filter(filterEmptyTextNodes)
          .map((item) => blockToSolid(item, options))
          .join('\n')}
        ${addWrapper ? '</>' : ''})
    }

    export default ${json.name};
  `;

    // HACK: for some reason we are generating `state.state.foo` instead of `state.foo`
    // need a full fix, but this unblocks a lot in the short term
    str = str.replace(/state\.state\./g, 'state.');

    if (options.plugins) {
      str = runPreCodePlugins(str, options.plugins);
    }
    if (options.prettier !== false) {
      str = format(str, {
        parser: 'typescript',
        plugins: [require('prettier/parser-typescript')],
      });
    }
    if (options.plugins) {
      str = runPostCodePlugins(str, options.plugins);
    }
    return str;
  };
