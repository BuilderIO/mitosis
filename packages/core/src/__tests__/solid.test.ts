import { componentToSolid } from '../generators/solid';
import { parseJsx } from '../parsers/jsx';
const basic = require('./data/basic.raw');
const submitButtonBlock = require('./data/blocks/submit-button.raw');
const inputBlock = require('./data/blocks/input.raw');
const selectBlock = require('./data/blocks/select.raw');
const formBlock = require('./data/blocks/form.raw');
const button = require('./data/blocks/button.raw');
const textarea = require('./data/blocks/textarea.raw');

describe('Solid', () => {
  test('Basic', () => {
    const json = parseJsx(basic);
    const output = componentToSolid(json);
    expect(output).toMatchSnapshot();
  });

  test('Input block', () => {
    const json = parseJsx(inputBlock);
    const output = componentToSolid(json);
    expect(output).toMatchSnapshot();
  });

  test('Submit button block', () => {
    const json = parseJsx(submitButtonBlock);
    const output = componentToSolid(json);
    expect(output).toMatchSnapshot();
  });

  test('Select block', () => {
    const json = parseJsx(selectBlock);
    const output = componentToSolid(json);
    expect(output).toMatchSnapshot();
  });

  test('Form block', () => {
    const json = parseJsx(formBlock);
    const output = componentToSolid(json);
    expect(output).toMatchSnapshot();
  });

  test('Button', () => {
    const json = parseJsx(button);
    const output = componentToSolid(json);
    expect(output).toMatchSnapshot();
  });

  test('Textarea', () => {
    const json = parseJsx(textarea);
    const output = componentToSolid(json);
    expect(output).toMatchSnapshot();
  });
});
