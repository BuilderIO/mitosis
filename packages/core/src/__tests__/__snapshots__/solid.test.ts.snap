// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Solid Basic 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";

export default function MyComponent() {
  const state = createMutable({ name: \\"Steve\\" });

  return (
    <div>
      <input
        value={state.name}
        onChange={(event) => (state.name = event.target.value)}
      />
      Hello! I can run in React, Vue, Solid, or Liquid!
    </div>
  );
}
"
`;

exports[`Solid Button 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";

export default function MyComponent() {
  const state = createMutable({});

  return (
    <Fragment>
      <Show when={props.link}>
        <a
          {...props.attributes}
          href={props.link}
          target={props.openLinkInNewTab ? \\"_blank\\" : undefined}
        >
          {props.text}
        </a>
      </Show>

      <Show when={!props.link}>
        <button {...props.attributes} type=\\"button\\">
          {props.text}
        </button>
      </Show>
    </Fragment>
  );
}
"
`;

exports[`Solid Form block 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";
import { css } from \\"solid-styled-components\\";

import { BuilderBlockComponent as BuilderBlock } from \\"@fake\\";
import { Builder, builder } from \\"@builder.io/sdk\\";
import { BuilderBlocks } from \\"@fake\\";
import { set } from \\"@fake\\";
import { get } from \\"@fake\\";

export default function MyComponent() {
  const state = createMutable({
    state: \\"unsubmitted\\",
    responseData: null,
    formErrorMessage: \\"\\",
    get submissionState() {
      return (Builder.isEditing && props.previewState) || state.state;
    },
    onSubmit(
      event: Event & {
        currentTarget: HTMLFormElement;
      }
    ) {
      const sendWithJs =
        props.sendWithJs || props.sendSubmissionsTo === \\"email\\";

      if (props.sendSubmissionsTo === \\"zapier\\") {
        event.preventDefault();
      } else if (sendWithJs) {
        if (!(props.action || props.sendSubmissionsTo === \\"email\\")) {
          event.preventDefault();
          return;
        }

        event.preventDefault();
        const el = event.currentTarget;
        const headers = props.customHeaders || {};
        let body: any;
        const formData = new FormData(el); // TODO: maybe support null

        const formPairs: {
          key: string;
          value: File | boolean | number | string | FileList;
        }[] = Array.from(
          event.currentTarget.querySelectorAll(\\"input,select,textarea\\")
        )
          .filter((el) => !!(el as HTMLInputElement).name)
          .map((el) => {
            let value: any;
            const key = (el as HTMLImageElement).name;

            if (el instanceof HTMLInputElement) {
              if (el.type === \\"radio\\") {
                if (el.checked) {
                  value = el.name;
                  return {
                    key,
                    value,
                  };
                }
              } else if (el.type === \\"checkbox\\") {
                value = el.checked;
              } else if (el.type === \\"number\\" || el.type === \\"range\\") {
                const num = el.valueAsNumber;

                if (!isNaN(num)) {
                  value = num;
                }
              } else if (el.type === \\"file\\") {
                // TODO: one vs multiple files
                value = el.files;
              } else {
                value = el.value;
              }
            } else {
              value = (el as HTMLInputElement).value;
            }

            return {
              key,
              value,
            };
          });
        let contentType = props.contentType;

        if (props.sendSubmissionsTo === \\"email\\") {
          contentType = \\"multipart/form-data\\";
        }

        Array.from(formPairs).forEach(({ value }) => {
          if (
            value instanceof File ||
            (Array.isArray(value) && value[0] instanceof File) ||
            value instanceof FileList
          ) {
            contentType = \\"multipart/form-data\\";
          }
        }); // TODO: send as urlEncoded or multipart by default
        // because of ease of use and reliability in browser API
        // for encoding the form?

        if (contentType !== \\"application/json\\") {
          body = formData;
        } else {
          // Json
          const json = {};
          Array.from(formPairs).forEach(({ value, key }) => {
            set(json, key, value);
          });
          body = JSON.stringify(json);
        }

        if (contentType && contentType !== \\"multipart/form-data\\") {
          if (
            // Zapier doesn't allow content-type header to be sent from browsers
            !(sendWithJs && props.action?.includes(\\"zapier.com\\"))
          ) {
            headers[\\"content-type\\"] = contentType;
          }
        }

        const presubmitEvent = new CustomEvent(\\"presubmit\\", {
          detail: {
            body,
          },
        });

        if (formRef) {
          formRef.dispatchEvent(presubmitEvent);

          if (presubmitEvent.defaultPrevented) {
            return;
          }
        }

        state.state = \\"sending\\";
        const formUrl = \`\${
          builder.env === \\"dev\\" ? \\"http://localhost:5000\\" : \\"https://builder.io\\"
        }/api/v1/form-submit?apiKey=\${builder.apiKey}&to=\${btoa(
          props.sendSubmissionsToEmail || \\"\\"
        )}&name=\${encodeURIComponent(props.name || \\"\\")}\`;
        fetch(
          props.sendSubmissionsTo === \\"email\\" ? formUrl : props.action!, // TODO: throw error if no action URL
          {
            body,
            headers,
            method: props.method || \\"post\\",
          }
        ).then(
          async (res) => {
            let body;
            const contentType = res.headers.get(\\"content-type\\");

            if (contentType && contentType.indexOf(\\"application/json\\") !== -1) {
              body = await res.json();
            } else {
              body = await res.text();
            }

            if (!res.ok && props.errorMessagePath) {
              // TODO: allow supplying an error formatter function
              let message = get(body, props.errorMessagePath);

              if (message) {
                if (typeof message !== \\"string\\") {
                  // TODO: ideally convert json to yaml so it woul dbe like
                  // error: - email has been taken
                  message = JSON.stringify(message);
                }

                state.formErrorMessage = message;
              }
            }

            state.responseData = body;
            state.state = res.ok ? \\"success\\" : \\"error\\";

            if (res.ok) {
              const submitSuccessEvent = new CustomEvent(\\"submit:success\\", {
                detail: {
                  res,
                  body,
                },
              });

              if (formRef) {
                formRef.dispatchEvent(submitSuccessEvent);

                if (submitSuccessEvent.defaultPrevented) {
                  return;
                } // TODO: option to turn this on/off?

                if (props.resetFormOnSubmit !== false) {
                  formRef.reset();
                }
              } // TODO: client side route event first that can be preventDefaulted

              if (props.successUrl) {
                if (formRef) {
                  const event = new CustomEvent(\\"route\\", {
                    detail: {
                      url: props.successUrl,
                    },
                  });
                  formRef.dispatchEvent(event);

                  if (!event.defaultPrevented) {
                    location.href = props.successUrl;
                  }
                } else {
                  location.href = props.successUrl;
                }
              }
            }
          },
          (err) => {
            const submitErrorEvent = new CustomEvent(\\"submit:error\\", {
              detail: {
                error: err,
              },
            });

            if (formRef) {
              formRef.dispatchEvent(submitErrorEvent);

              if (submitErrorEvent.defaultPrevented) {
                return;
              }
            }

            state.responseData = err;
            state.state = \\"error\\";
          }
        );
      }
    },
  });

  const formRef = useRef();

  return (
    <form
      {...props.attributes}
      validate={props.validate}
      ref={formRef}
      action={!props.sendWithJs && props.action}
      method={props.method}
      name={props.name}
      onSubmit={(event) => state.onSubmit(event)}
    >
      <Show when={props.builderBlock && props.builderBlock.children}>
        <For each={props.builderBlock?.children}>
          <BuilderBlockComponent
            key={block.id}
            block={block}
          ></BuilderBlockComponent>
        </For>
      </Show>

      <Show when={state.submissionState === \\"error\\"}>
        <BuilderBlocks
          dataPath=\\"errorMessage\\"
          blocks={props.errorMessage!}
        ></BuilderBlocks>
      </Show>

      <Show when={state.submissionState === \\"sending\\"}>
        <BuilderBlocks
          dataPath=\\"sendingMessage\\"
          blocks={props.sendingMessage!}
        ></BuilderBlocks>
      </Show>

      <Show when={state.submissionState === \\"error\\" && state.responseData}>
        <pre
          class={css({
            padding: \\"10px\\",
            color: \\"red\\",
            textAlign: \\"center\\",
          })}
        >
          {JSON.stringify(state.responseData, null, 2)}
        </pre>
      </Show>

      <Show when={state.submissionState === \\"success\\"}>
        <BuilderBlocks
          dataPath=\\"successMessage\\"
          blocks={props.successMessage!}
        ></BuilderBlocks>
      </Show>
    </form>
  );
}
"
`;

exports[`Solid Img 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";

import { Builder } from \\"@builder.io/sdk\\";

export default function MyComponent() {
  const state = createMutable({});

  return (
    <img
      {...props.attributes}
      style={{
        objectFit: props.backgroundSize || \\"cover\\",
        objectPosition: props.backgroundPosition || \\"center\\",
      }}
      key={(Builder.isEditing && props.imgSrc) || \\"default-key\\"}
      alt={props.altText}
      src={props.imgSrc}
    />
  );
}
"
`;

exports[`Solid Input block 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";

import { Builder } from \\"@builder.io/sdk\\";

export default function MyComponent() {
  const state = createMutable({});

  return (
    <input
      {...props.attributes}
      key={
        Builder.isEditing && props.defaultValue
          ? props.defaultValue
          : \\"default-key\\"
      }
      placeholder={props.placeholder}
      type={props.type}
      name={props.name}
      value={props.value}
      defaultValue={props.defaultValue}
      required={props.required}
    />
  );
}
"
`;

exports[`Solid Section 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";

export default function MyComponent() {
  const state = createMutable({});

  return (
    <section
      {...props.attributes}
      style={
        props.maxWidth && typeof props.maxWidth === \\"number\\"
          ? {
              maxWidth: props.maxWidth,
            }
          : undefined
      }
    >
      {props.children}
    </section>
  );
}
"
`;

exports[`Solid Select block 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";

import { Builder } from \\"@builder.io/sdk\\";

export default function MyComponent() {
  const state = createMutable({});

  return (
    <select
      {...props.attributes}
      value={props.value}
      key={
        Builder.isEditing && props.defaultValue
          ? props.defaultValue
          : \\"default-key\\"
      }
      defaultValue={props.defaultValue}
      name={props.name}
    >
      <For each={props.options}>
        <option value={option.value}>{option.name || option.value}</option>
      </For>
    </select>
  );
}
"
`;

exports[`Solid Submit button block 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";

export default function MyComponent() {
  const state = createMutable({});

  return (
    <button {...props.attributes} type=\\"submit\\">
      {props.text}
    </button>
  );
}
"
`;

exports[`Solid Text 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";

import { Builder } from \\"@builder.io/sdk\\";

export default function MyComponent() {
  const state = createMutable({});

  return (
    <div
      contentEditable={allowEditingText || undefined}
      innerHTML={props.text || props.content || \\"\\"}
    ></div>
  );
}
"
`;

exports[`Solid Textarea 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";

export default function MyComponent() {
  const state = createMutable({});

  return (
    <textarea
      {...props.attributes}
      placeholder={props.placeholder}
      name={props.name}
      value={props.value}
      defaultValue={props.defaultValue}
    ></textarea>
  );
}
"
`;

exports[`Solid Video 1`] = `
"import { createMutable, Show, For } from \\"solid-js\\";

export default function MyComponent() {
  const state = createMutable({});

  return (
    <video
      {...props.attributes}
      preload=\\"none\\"
      style={{
        width: \\"100%\\",
        height: \\"100%\\",
        ...props.attributes?.style,
        objectFit: props.fit,
        objectPosition: props.position,
        // Hack to get object fit to work as expected and
        // not have the video overflow
        borderRadius: 1,
      }}
      key={props.video || \\"no-src\\"}
      poster={props.posterImage}
      autoPlay={props.autoPlay}
      muted={props.muted}
      controls={props.controls}
      loop={props.loop}
    ></video>
  );
}
"
`;
